core/modules/replica/HttpModule.h

```diff
@@ -0,0 +1,140 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_HTTPMODULE_H
+#define LSST_QSERV_HTTPMODULE_H
+
+// System headers
+#include <memory>
+#include <string>
+
+// Third party headers
+#include "nlohmann/json.hpp"
+
+// Qserv headers
+#include "qhttp/Server.h"
+#include "replica/EventLogger.h"
+
+// This header declarations
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+ * Class HttpModule is a base class for requests processing modules an HTTP server
+ * built into the Master Replication Controller.
+ */
+class HttpModule:
+        public EventLogger,
+        public std::enable_shared_from_this<HttpModule> {
+public:
+
+    typedef std::shared_ptr<HttpModule> Ptr;
+
+    HttpModule() = delete;
+    HttpModule(HttpModule const&) = delete;
+    HttpModule& operator=(HttpModule const&) = delete;
+
+    virtual ~HttpModule() = default;
+
+    /**
+     * Invokes a subclass-specific request processing. Also does an optional
+     * (if any) processing of exceptions thrown by the subclass-specific
+     * processors. These error conditions will be reported to as errors to callers.
+     * 
+     * @param req  the HTTP request
+     * @param resp  the HTTP response channel
+     * @param subModuleName  this optional parameter allows modules to have
+     * multiple sub-modules. A value of this parameter will be forwarded to
+     * the subclass-specific implementation of the pure virtual method
+     * HttpModule::executeImpl().
+     */
+    void execute(qhttp::Request::Ptr const& req,
```

Comment:
should this be marked either `virtual` or `final`? I guess I can't tell if this is supposed to be overridden or if it calls a `do` method that subclasses are supposed to override. Maybe I'll find out later... :-)

Reply:
Ok I looked it up and you can't do that. Too bad! The theory is that since it wasn't marked `virtual` then you're supposed to know that it's intentional that you should not override it. I think that places too much faith on the future readers to trust the past authors of intentional omission, but I guess that's the way it goes. Maybe if it's clearly explained that this calls function that subclasses should override?

Reply:
Okay, I have rewritten the documentation of the method. Hopefully, this will help.